### ISSUE TITLE
Performance Degradation 3.6.00 to 3.6.01

### ISSUE THREAD

**Issue by wjhorne** [kokkos/kokkos#6533]
**Describe the bug**
The SPARC team found a pretty serious drop in performance on Nvidia hardware due to recent updates in Kokkos. @vbrunini found one change in particular in the update from 3.6.00 to 3.6.01 that is a direct culprit behind the drop listed below. Can we get some help addressing this?

**Please also include the following items to support reproducing the bug**
1. compilers (with versions)
NVCC 11.2
2. Kokkos release or commit used (i.e., the sha1 number)
3.6.01
3. platform and backend
CUDA
4. Code change responsible for drop
```
diff --git a/packages/kokkos/core/src/impl/Kokkos_ViewTracker.hpp b/packages/kokkos/core/src/impl/Kokkos_ViewTracker.hpp
index 972b1b6d9af..fe3651886bd 100644
--- a/packages/kokkos/core/src/impl/Kokkos_ViewTracker.hpp
+++ b/packages/kokkos/core/src/impl/Kokkos_ViewTracker.hpp
@@ -91,7 +91,6 @@ struct ViewTracker {

   template <class RT, class... RP>
   KOKKOS_INLINE_FUNCTION void assign(const View<RT, RP...>& vt) noexcept {
-    if (this == reinterpret_cast<const ViewTracker*>(&vt.m_track)) return;
     KOKKOS_IF_ON_HOST((
         if (view_traits::is_managed && Kokkos::Impl::SharedAllocationRecord<
                                            void, void>::tracking_enabled()) {
@@ -103,7 +102,6 @@ struct ViewTracker {

   KOKKOS_INLINE_FUNCTION ViewTracker& operator=(
       const ViewTracker& rhs) noexcept {
-    if (this == &rhs) return *this;
     KOKKOS_IF_ON_HOST((
         if (view_traits::is_managed && Kokkos::Impl::SharedAllocationRecord<
                                            void, void>::tracking_enabled()) {
```

**Comment by dalg24** [comment]
(For reference) this change was introduced in #5024

**Comment by dalg24** [comment]
Please confirm with you see that performance degradation when undoing the change on the develop branch.

**Comment by vbrunini** [comment]
Yes it looks like current develop still has the regression and commenting out the self-assignment checks recovers the performance.

**Comment by masterleinad** [comment]
@vbrunini Would you have a reproducer for the performance degradation? How big is the slowdown?

**Comment by vbrunini** [comment]
This ifpack2 example reproduces the issue: https://github.com/trilinos/Trilinos/blob/master/packages/ifpack2/example/BlockTriDi.cpp 

Running `jsrun -n 4 -g 1 -M "-gpu" packages/ifpack2/example/Ifpack2_BlockTriDiagonalSolver.exe --matrixType=Laplace3D --blockSize=5 --nx=64 --ny=64 --nz=128 --tol=0 --numIters=20 --numRepeats=50 --withStackedTimer` on our power9/V100 machine I see:

With self-assignment check:
```
BlockTriDiagonalSolver: 1.82134 [1] {min=1.81389, max=1.82799, std dev=0.00577539} <1, 2, 1>
|   Preconditioner setup: 0.0433162 - 2.37826% [1] {min=0.0412015, max=0.0446555, std dev=0.00148679} <1, 0, 3>
|   Preconditioner compute: 0.0883114 - 4.84871% [50] {min=0.087768, max=0.0886765, std dev=0.000384927} <1, 0, 3>
|   Solve: 1.57542 - 86.4982% [50] {min=1.57101, max=1.57733, std dev=0.00298764} <1, 0, 3>
|   Remainder: 0.114285 - 6.27481%
End Result: TEST PASSED
```

Without self-assignment check:
```
BlockTriDiagonalSolver: 1.33279 [1] {min=1.32596, max=1.33522, std dev=0.00455889} <1, 0, 3>
|   Preconditioner setup: 0.0443767 - 3.3296% [1] {min=0.0421491, max=0.0457239, std dev=0.00157572} <1, 1, 2>
|   Preconditioner compute: 0.0821505 - 6.16379% [50] {min=0.0818588, max=0.0826194, std dev=0.00033174} <2, 1, 1>
|   Solve: 1.09883 - 82.4459% [50] {min=1.09473, max=1.10102, std dev=0.00296268} <1, 1, 2>
|   Remainder: 0.107432 - 8.06065%
```

**Comment by crtrott** [comment]
Hm are any of the kernels getting slower? Or is it more time on the host. And are we sure you get the right answer without leaking memory or anything like that? 

One test: what if you move that self assignment guard into the KOKKOS_IF_ON_HOST? What does that do to performance.

**Comment by vbrunini** [comment]
Yes, this kernel is where the slowdown is: https://github.com/trilinos/Trilinos/blob/d1f042da58a3e1de8bb198a5bfc3821608db625b/packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp#L3454

and all of our tests pass without the self-assignment check. At one point debugging a separate issue we put an assert in the View assignments and confirmed that we never self-assign any Views.

I haven't checked guarding with KOKKOS_IF_ON_HOST specifically, but I'm pretty confident it would resolve the performance issue. I'm currently testing out a patch to ifpack2 that removes the subview usage from the function linked above which seems to recover most of the performance on the example, but I want to test that on some full problems as well.

@kliegeois @srajama1 FYI

**Comment by crtrott** [comment]
Hm interesting, it should already be compile time unmanaged. We should check why this stuff is at all invoked with tracking.

**Comment by crtrott** [comment]
Just to make that clear the ifpack two fix used there is definitely illegal, and may in the future break in debug mode. Its a precondition violation to construct a Kokkos View from an invalid pointer.

**Comment by vbrunini** [comment]
I confirmed that moving the self assignment guard inside the KOKKOS_IF_ON_HOST resolves the performance issue.

**Comment by ajpowelsnl** [comment]
Hi @vbrunini - is the performance issue fully resolved?

**Comment by ajpowelsnl** [comment]
**TODO**
- Test with current release
- Testing w/ [conditions above](https://github.com/kokkos/kokkos/issues/6533#issue-1958009863)

**Comment by ajpowelsnl** [comment]
@vbrunini -- can you provide the Trilinos build recipe for the [reproducer](https://github.com/kokkos/kokkos/issues/6533#issuecomment-1777469839)?

**Comment by vbrunini** [comment]
I'm out of the office for a couple of weeks, but @kliegeois should be able to help with building that Ifpack2 example. You'll likely have to checkout an old version of Trilinos to test though, we removed the subview calls that were causing the slowdown after the Kokkos change, and the kernel in question has been further refactored since then so Trilinos/develop wouldn't reproduce it anymore.

**Comment by ajpowelsnl** [comment]
Thanks, @vbrunini -- given your changes, is this issue still relevant / something for us to fix?
